src/v/s3/client.cc

```diff
@@ -827,6 +827,77 @@ ss::future<> client::delete_object(
       });
 }
 
+static auto iobuf_to_delete_objects_result(iobuf&& buf) {
+    auto root = iobuf_to_ptree(std::move(buf));
+    auto result = client::delete_objects_result{};
+    try {
+        for (auto const& [tag, value] : root.get_child("DeleteResult")) {
+            if (tag != "Error") {
+                continue;
+            }
+            auto code = value.get<std::string>(
+              "Code", "[no error code present]");
+            auto key = value.get<std::string>("Key", "[no key present]");
+            auto message = value.get<std::string>(
+              "Message", "[no error message present]");
+            auto version_id = value.get<std::string>(
+              "VersionId", "[no version id present]");
+            vlog(
+              s3_log.trace,
+              R"log(delete_objects_result::undeleted_keys Code:"{}" Key:"{}" Message:"{}" VersionId:"{}")log",
+              code,
+              key,
+              message,
+              version_id);
+            if (value.find("Key") == value.not_found()) {
+                vlog(
+                  s3_log.warn,
+                  "an DeleteResult.Error does not contain the Key tag");
+                continue;
+            }
+            result.undeleted_keys.push_back({
+              s3::object_key{value.get<std::string>("Key")},
+              value.get<std::string>("Code", "[no error code present]"),
+            });
+        }
+    } catch (...) {
+        vlog(s3_log.error, "!!error parse result {}", std::current_exception());
+        throw;
+    }
+    return result;
+}
+
+auto client::delete_objects(
+  bucket_name const& bucket,
+  std::span<const object_key> keys,
+  ss::lowres_clock::duration timeout) -> ss::future<delete_objects_result> {
+    auto request = _requestor.make_delete_objects_request(bucket, keys);
+    if (!request) {
+        return ss::make_exception_future<delete_objects_result>(
+          std::system_error(request.error()));
+    }
+    auto& [header, body] = request.value();
+    vlog(s3_log.trace, "send https request:\n{}", header);
+
+    return ss::do_with(
+             std::move(body),
+             [&_client = _client, header = std::move(header), timeout](
+               auto& to_delete) mutable {
+                 return _client.request(std::move(header), to_delete, timeout);
+             })
+      .then([](http::client::response_stream_ref const& response) {
+          return drain_response_stream(response).then([response](iobuf&& res) {
+              auto status = response->get_headers().result();
+              if (status != boost::beast::http::status::ok) {
```

Comment:
Sometimes S3 gives you an http error code without the body (like when you're throttled heavily). I think the code will work in this case but the `undeleted_keys` field will have no data despite the fact that the keys are not deleted at all. Maybe you can just always set `undeleted_keys`  to the list of keys passed to the `parse_rest_error_response` function? It's safe to try to delete something twice in S3.

Reply:
not sure I follow, because undeleted_keys are generated only in the boost::beast::http::status::ok path.

for good measure, I added a testcase to check that an empty body is translated to a slow_down error code